qiskit/circuit/library/phase_oracle.py

```diff
@@ -0,0 +1,120 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Phase Oracle object."""
+
+from qiskit.circuit import QuantumCircuit, QuantumRegister
+from qiskit.circuit.classicalfunction.boolean_expression import BooleanExpression
+
+
+class PhaseOracle(QuantumCircuit):
+    r"""Phase Oracle.
+
+    The Phase Oracle object constructs circuits for any arbitrary
+    input logical expressions. A logical expression is composed of logical operators
+    `&` (`AND`), `|` (`OR`), `~` (`NOT`), and `^` (`XOR`).
+    as well as symbols for literals (variables).
+    For example, `'a & b'`, and `(v0 | ~v1) & (~v2 & v3)`
+    are both valid string representation of boolean logical expressions.
+
+    For convenience, this oracle, in addition to parsing arbitrary logical expressions,
+    also supports input strings in the `DIMACS CNF format
+    <http://www.satcompetition.org/2009/format-benchmarks2009.html>`__,
+    which is the standard format for specifying SATisfiability (SAT) problem instances in
+    `Conjunctive Normal Form (CNF) <https://en.wikipedia.org/wiki/Conjunctive_normal_form>`__,
+    which is a conjunction of one or more clauses, where a clause is a disjunction of one
+    or more literals. See :meth:`qiskit.circuit.library.phase_oracle.PhaseOracle.from_dimacs_file`.
+    """
+
+    def __init__(self, expression: str) -> None:  # pylint: disable=super-init-not-called
```

Comment:
This structure with `_build_from_boolean_expression` seems a bit weirdly split to me. Couldn't we just have all of it in the `__init__`? Then we also wouldn't have to store `self.state_qubits` 🙂

Reply:
How the constructor call would look like?

Reply:
hmm after trying to implement it, storing `boolean_expression` does make a lot of sense since you need it for the `evaluate` method. But we can still avoid storing `state_qubits` by rewriting `_build_from_boolean_expression` a bit:
```python
def _build_from_boolean_expression(self):
    qr_state = QuantumRegister(self.boolean_expression.num_qubits - 1, 'state')
    # rest as before
```
I'd prefer storing as little extra info as necessary, otherwise users might expect `state_qubits` available on other circuits.

Reply:
oh! I thought `state_qubits` was part of the `PhaseOracle` API. Removed in 261ccc097